commit bdd2df03a3372347904deb970cef1dbd9865b522
Author: Alban Crequy <alban.crequy@collabora.co.uk>
Date:   Tue Nov 3 18:23:02 2009 +0000

    Fix bug #139065 gnome-vfs-daemon subscribes to HAL D-Bus signals too broadly

diff --git a/libgnomevfs/gnome-vfs-hal-mounts.c b/libgnomevfs/gnome-vfs-hal-mounts.c
index 61e7147..5e99d2d 100644
--- a/libgnomevfs/gnome-vfs-hal-mounts.c
+++ b/libgnomevfs/gnome-vfs-hal-mounts.c
@@ -1124,6 +1124,9 @@ _hal_update_all (GnomeVFSVolumeMonitorDaemon *volume_monitor_daemon)
 #ifdef HAL_SHOW_DEBUG
 						g_debug ("  volume = '%s'", volumes[j]);
 #endif
+						libhal_device_add_property_watch (volume_monitor_daemon->hal_ctx,
+										  volumes[j], NULL);
+
 						volume = libhal_volume_from_udi (volume_monitor_daemon->hal_ctx, 
 										 volumes[j]);
 
@@ -1182,6 +1185,9 @@ _hal_device_added (LibHalContext *hal_ctx,
 		LibHalVolume *volume;
 		DBusError error;
 
+		libhal_device_add_property_watch (volume_monitor_daemon->hal_ctx,
+			udi, NULL);
+
 		drive = NULL;
 		volume = NULL;
 
@@ -1271,6 +1277,16 @@ _hal_device_removed (LibHalContext *hal_ctx, const char *udi)
 
 	hal_drive_udi = NULL;
 
+	/* We don't know whether the removed device is a volume. If it is not
+	 * a volume, libhal_device_remove_property_watch() will do nothing. It
+	 * cause a round-trip to the dbus-daemon but it is harmless because
+	 * the event "device removed" happens seldom. We cannot use the check
+	 * volume != NULL because a volume is added in the monitor only if it
+	 * is mounted.
+	 *  */
+	libhal_device_remove_property_watch
+		(volume_monitor_daemon->hal_ctx, udi, NULL);
+
 	if (volume != NULL) {
 #ifdef HAL_SHOW_DEBUG
 		g_debug ("Removing GnomeVFSVolume for device path %s", volume->priv->device_path);
@@ -1478,12 +1494,6 @@ _gnome_vfs_hal_mounts_init (GnomeVFSVolumeMonitorDaemon *volume_monitor_daemon)
 	libhal_ctx_set_user_data (volume_monitor_daemon->hal_ctx,
 				  hal_userdata);
 
-	/* Simply watch all property changes instead of dynamically
-	 * adding/removing match rules bus-side to only match certain
-	 * objects...
-	 */
-	libhal_device_property_watch_all (volume_monitor_daemon->hal_ctx, &error);
-
 	/* add drives/volumes from HAL */
 	_hal_update_all (volume_monitor_daemon);
 
